# Cases

## PBMC example

### Marker gene heatmap

```{r marker-gene-heatmap, fig.width=10, fig.height=7}
library(yulab.utils)
pload(dplyr)
pload(ggplot2)
pload(ggfun)
pload(ggtree)
pload(aplot)
pload(Seurat)


# celltype <- c("Naive CD4 T", "B", "NK")
# id <- pbmc@active.ident
# md = data.frame(cell=names(id), group=id)
# md <- md[md$group %in% celltype, ]
# md$group <- as.character(md$group)
# md <- lapply(split(md, md$group), function(x) x[1:30,]) |> rbindlist()
#
# pbmc <- pbmc[, md$cell]

pbmc <- qs::qread("data/pbmc-subset.qs")

m <- FindAllMarkers(pbmc, only.pos=TRUE)

top10 <- m %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 2) %>%
    slice_head(n = 10) 

pbmc <- ScaleData(pbmc, features = rownames(pbmc))
x  <- pbmc[['RNA']]$scale.data[top10$gene,]

y <- mat2df(x) -> y
y$gene <- rownames(x)[y$row]
y$cell <- colnames(x)[y$col]

p <- ggplot(y, aes(cell, gene, fill = value)) + 
    geom_tile()+ scale_fill_viridis_c(name = "Gene Expression") +
    theme_noxaxis() +
    scale_y_discrete(position = "right") +
    xlab(NULL) +
    ylab(NULL)

gene_cls <- hclust(dist((x)))
cell_cls <- hclust(dist(t((x)))) 
gene_tree <- ggtree(gene_cls, branch.length = 'none')
cell_tree = ggtree(cell_cls, branch.length = 'none') + layout_dendrogram()


id <- pbmc@active.ident
md = data.frame(cell=names(id), group=id)
p_cell_type <- ggplot(md, aes(cell, y=1, fill = group)) + 
    geom_tile() + 
    scale_fill_brewer(palette="Set1", name = 'Cell Type') +
    theme_nothing()

g <- p |> insert_left(gene_tree, width = .2) |> 
    insert_top(p_cell_type, height = .02) |> 
    insert_top(cell_tree, height = .1)

g
```

### Degree of genes in the PPI network

```{r ppi}
pload(igraph)
pload(clusterProfiler)

ppi <- getPPI(top10$gene, output='igraph', taxID='9606') 
deg <- stack(degree(ppi, v = V(ppi))) |> setNames(c("degree", "gene"))

p_ppi <- ggplot(deg, aes(1, gene, fill=degree, size=degree)) + geom_point(shape=21) + 
    scale_fill_gradientn(colors = hcl.colors(20, "RdYlGn")) + theme_nothing()
```

### Genes in pathways

```{r pcp, fig.width=16, fig.height=7}
pload(tidytree)
pload(msigdbr)

node <- c(35, 34, 32)
genes <- lapply(node, function(n) offspring(gene_tree, n, tiponly=T)$label) |>
    setNames(c("CD4", "NK", "B"))

pload(clusterProfiler)
gene_set <- msigdbr(species="human", category="C2")
names(gene_set)
cc <- compareCluster(genes, enricher, TERM2GENE=gene_set[, c("gs_name", "gene_symbol")])
res <- cc@compareClusterResult |>
    group_by(Cluster) |> 
    slice_head(n = 5) |> 
    as.data.frame()

gene2path <- lapply(res$geneID, function(x) unlist(strsplit(x, split="/"))) |>
    setNames(res$Description) |>
    stack() |>
    setNames(c("gene", "pathway"))

pcp <- ggplot(gene2path, aes(pathway, gene))  +
    geom_point(size=5, shape=21, fill = "steelblue") +
    theme_minimal() + 
    theme_noyaxis() +
    theme(axis.text.x = element_text(angle=30, hjust=1)) + 
    xlab(NULL) +
    ylab(NULL)

gg <- g |> insert_right(p_ppi, width=.05) |>
    insert_right(pcp, width=.5) 
gg
```



### All in one

```{r all, fig.width=16, fig.height=14}
pp <- plot_list(cell_tree, p_cell_type, heights=c(1, .5))
pp <- plot_list(gene_tree, pp, widths=c(.3, 1))
pp <- plot_list(p, pp, p_ppi, pcp, widths=c(1, .8, .05, .6))
pp <- pp + patchwork::plot_annotation(tag_levels='a')

final_plot <- plot_list(pp, gg, ncol=1, tag_levels='A', heights=c(1, 1.2))
final_plot
```

## Oncoplot example


```{r onco, fig.width=16, fig.height=7}
pload(aplotExtra)
pload(RTCGA.mRNA)
pload(RTCGA)

laml.maf <- system.file("extdata", "tcga_laml.maf.gz", package = "maftools")
laml.clin <- system.file('extdata', 'tcga_laml_annot.tsv', package = 'maftools')
laml <- maftools::read.maf(maf = laml.maf, clinicalData = laml.clin)
onco <- oncoplot(maf = laml, genes = 20)

plot_tcga_expr <- function(mRNA, genes, name = "Gene Expression") {
    d = expressionsTCGA(mRNA, extract.cols = genes)

    dd = gather(d, gene, expression, -c(1,2))
    
    ggplot(dd, aes(expression, gene, fill=stat(x))) + 
        ggridges::geom_density_ridges_gradient() +
        scale_fill_viridis_c(option="C", name = name) +
        theme_minimal() +
        theme_noyaxis() + 
        xlab(NULL) +
        ylab(NULL) +
        theme(legend.position='bottom')
}

onco_genes <- aplotExtra:::get_oncoplot_genes(laml)
brca <- plot_tcga_expr(BRCA.mRNA, onco_genes, "Gene Expression in TCGA\nbreast cancer patients")
ov <- plot_tcga_expr(OV.mRNA, onco_genes, "Gene Expression in TCGA\novarian cancer patients")

op <- insert_right(onco, brca, width = .6) |>
    insert_right(ov, width=.6)
op
```

## funky heatmap


```{r funky, fig.width=8, fig.height=8}
library(aplotExtra)
library(tidyverse)
data("mtcars")


d <- yulab.utils::scale_range(mtcars) |>
    rownames_to_column("id") |>
    arrange(desc(mpg))


g1 <- funky_text(d)
g2 <- funky_bar(d, 2) + scale_fill_gradient(low = "#CC4C02", high = "#FFFFE5")
g3 <- funky_bar(d, 3) + scale_fill_gradient(low = "steelblue", high = "firebrick") 
g4 <- funky_point(d, 4:7) + scale_fill_gradient(low = "#CC4C02", high = "#FFFFE5")
g5 <- funky_point(d, 8:12) + scale_fill_gradient(low = "#08519C", high = "#F7FBFF") 


#funky_heatmap(data=mtcars)
fp <- funky_heatmap(g1, g2, g4, g5, options=theme_stamp())
fp
```


## Epigenetics 
In this part, we will show case of epigenetics by using aplot to integrate the visualization
of hic data, ChIP-seq data and gene track. Hic data will be manipulated and visualized by 
`HiContacts` and `HiCExperiment`. ChIP-seq data will be manipulated and visualize by `ChIPseeker`.
Gene track will be visualized by `gggenes` by fetching gene information from txdb object.
aplot will show its power by combining these three parts of epigenetics result.

The purpose of this case is to show the different epigenetics characteristics of cell line 
GM12878(normal lymphoblastoid cell line) and cell line K562(chronic myeloid leukemia cell lines).
Hic data are derived from GM12878([GSM1551550](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1551550)) and K562([GSM1551618](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1551618)). 
ChIP-seq data have four parts: H3K27ac, H3K4me1, H3K4me3 and H3K36me3.
ChIP-seq Control data are derived from GM12878([GSM733742](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733742)) and K562([GSM733780](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733780)).
ChIP-seq H3K27ac data are derived from GM12878([GSM733771](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733771)) and K562([GSM733656](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733656)).
ChIP-seq H3K4me1 data are derived from GM12878([GSM733772](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733772)) and K562([GSM733692](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733692)).
ChIP-seq H3K4me3 data are derived from GM12878([GSM733708](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733708)) and K562([GSM733680](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733680)).
ChIP-seq H3K36me3 data are derived from GM12878([GSM733679](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733679)) and K562([GSM733714](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733714)).
We download the bam files and re-analyze it by using macs2 to get peak results and different peak results. 
In order to manifest the strength of aplot, we will show the usages of aplot in two parts.
combine data is the comparison result. The comparison result of Hic data are calculated by using `HiContacts::divide` to get
foldchange(K562/GM12878). The comparison result of ChIP-seq data are calculated by using the `bdgdiff` function of `macs2`.
split data is to exhibit the hic data and ChIP-seq data of two cell lines individually.

All analyzed result store in epigenetics_data.qs file.
This file have nine parts. `epigenetics_data[["H3K4me1_peak"]]` stores differential peak analysis result of H3K4me1. 
`epigenetics_data[["H3K4me3_peak"]]`, `epigenetics_data[["H3K27ac_peak"]]` and `epigenetics_data[["H3K36me3_peak"]]` are also differential peak analysis results.
`epigenetics_data[["GM12878_peak"]]` and `epigenetics_data[["K562_peak"]]` store the ChIP-seq data of GM12878 and K562.
`epigenetics_data[["GM12878_hic"]]` and `epigenetics_data[["K562_hic"]]` store the hic data.
`epigenetics_data[["div_hic"]]` store the compare result of GM12878 and K562 hic data.

Following code show how to preprocess Hic data.
```{r, eval=F}
# deal with hic data
# download GM12878.hic
# wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM1551550&format=file&file=GSM1551550%5FHIC001%5F30%2Ehic
# download K562.hic
# wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM1551618&format=file&file=GSM1551618%5FHIC069%5F30%2Ehic
GM12878_hic <- HiCExperiment::import("GM12878.hic",
                                    focus = "6:43600000-44100000",
                                    resolution = 5000)
K562_hic <- HiCExperiment::import("K562.hic",
                                  focus = "6:43600000-44100000",
                                  resolution = 5000)

epigenetics_data[["K562_hic"]] <- K562_hic
epigenetics_data[["GM12878_hic"]] <- GM12878_hic
epigenetics_data[["div_hic"]] <- HiContacts::divide("K562.hic", by = "GM12878.hic")

```

Following code show how to preprocess ChIP-seq data
```{sh,eval = F}
# GM12878 Control_rep2.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733742)
wget https://www.ncbi.nlm.nih.gov/geo/download/\?acc\=GSM733742\&format\=file\&file\=GSM733742%5Fhg19%5FwgEncodeBroadHistoneGm12878ControlStdAlnRep2%2Ebam

# GM12878 H3K4me1_rep2.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733772)
wget https://www.ncbi.nlm.nih.gov/geo/download/\?acc\=GSM733772\&format\=file\&file\=GSM733772%5Fhg19%5FwgEncodeBroadHistoneGm12878H3k4me1StdAlnRep2%2Ebam

# GM12878 H3K4me3_rep2.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733708)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733708&format=file&file=GSM733708%5Fhg19%5FwgEncodeBroadHistoneGm12878H3k4me3StdAlnRep2%2Ebam

# GM12878 H3K27ac_rep2.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733771)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733771&format=file&file=GSM733771%5Fhg19%5FwgEncodeBroadHistoneGm12878H3k27acStdAlnRep2%2Ebam

# GM12878 H3K36me3_rep2.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733679)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733679&format=file&file=GSM733679%5Fhg19%5FwgEncodeBroadHistoneGm12878H3k36me3StdAlnRep2%2Ebam

# GM12878 H3K4me1, output GM12878_H3K4me1_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep2.bam -t H3K4me1_rep2.bam -f BAM -g 2.7e9 -n GM12878_H3K4me1 -B --outdir . >macs2.log 2>&1 &

# GM12878 H3K4me3, output GM12878_H3K4me3_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep2.bam -t H3K4me3_rep2.bam -f BAM -g 2.7e9 -n GM12878_H3K4me3 -B --outdir . >macs2.log 2>&1 &

# GM12878 H3K27ac, output GM12878_H3K27ac_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep2.bam -t H3K27ac_rep2.bam -f BAM -g 2.7e9 -n GM12878_H3K27ac -B --outdir . >macs2.log 2>&1 &

# GM12878 H3K36me3, output GM12878_H3K36me3_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep2.bam -t H3K36me3_rep2.bam -f BAM -g 2.7e9 -n GM12878_H3K36me3 -B --outdir . >macs2.log 2>&1 &

# K562 Control_rep1.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733780)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733780&format=file&file=GSM733780%5Fhg19%5FwgEncodeBroadHistoneK562ControlStdAlnRep1%2Ebam

# K562 H3K4me1_rep1.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733692)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733692&format=file&file=GSM733692%5Fhg19%5FwgEncodeBroadHistoneK562H3k4me1StdAlnRep1%2Ebam

# K562 H3K4me3_rep1.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733680)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733680&format=file&file=GSM733680%5Fhg19%5FwgEncodeBroadHistoneK562H3k4me3StdAlnRep1%2Ebam

# K562 H3K27ac_rep1.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733656)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733656&format=file&file=GSM733656%5Fhg19%5FwgEncodeBroadHistoneK562H3k27acStdAlnRep1%2Ebam

# K562 H3K36me3_rep1.bam (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM733714)
wget https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM733714&format=file&file=GSM733714%5Fhg19%5FwgEncodeBroadHistoneK562H3k36me3StdAlnRep1%2Ebam

# K562 H3K4me1, output K562_H3K4me1_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep1.bam -t H3K4me1_rep1.bam -f BAM -g 2.7e9 -n K562_H3K4me1 -B --outdir . >macs2.log 2>&1 &

# K562 H3K4me3, output K562_H3K4me3_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep1.bam -t H3K4me3_rep1.bam -f BAM -g 2.7e9 -n K562_H3K4me3 -B --outdir . >macs2.log 2>&1 &

# K562 H3K27ac, output K562_H3K27ac_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep1.bam -t H3K27ac_rep1.bam -f BAM -g 2.7e9 -n K562_H3K27ac -B --outdir . >macs2.log 2>&1 &

# K562 H3K36me3, output K562_H3K36me3_peaks.narrowPeak
nohup macs2 callpeak -c Control_rep1.bam -t H3K36me3_rep1.bam -f BAM -g 2.7e9 -n K562_H3K36me3 -B --outdir . >macs2.log 2>&1 &

# --d1 --d2 is the sequence depth in control file(https://github.com/macs3-project/MACS/wiki/Call-differential-binding-events), user can cat .log file in /callpeak_result/xxxx/macs2.log to obtain.
# H3K4me1
# GM12878 control 6284864
# K562 control 18559220
nohup macs2 bdgdiff --t1 GM12878_H3K4me1_treat_pileup.bdg --c1 GM12878_H3K4me1_control_lambda.bdg --d1 6284864 --t2 K562_H3K4me1_treat_pileup.bdg --c2 K562_H3K4me1_control_lambda.bdg --d2 18559220 --outdir . -o H3K4me1_GM12878.bed H3K4me1_K562.bed nondiff_peak.bed 1>bdgdiff.log 2>&1 &

# H3K4me3
# GM12878 control 6284864
# K562 control 18559220
nohup macs2 bdgdiff --t1 GM12878_H3K4me3_treat_pileup.bdg --c1 GM12878_H3K4me3_control_lambda.bdg --d1 6284864 --t2 K562_H3K4me3_treat_pileup.bdg --c2 K562_H3K4me3_control_lambda.bdg --d2 18559220 --outdir . -o H3K4me3_GM12878.bed H3K4me3_K562.bed nondiff_peak.bed 1>bdgdiff.log 2>&1 &

# H3K27ac
# GM12878 control 6284864
# K562 control 18559220
nohup macs2 bdgdiff --t1 GM12878_H3K27ac_treat_pileup.bdg --c1 GM12878_H3K27ac_control_lambda.bdg --d1 6284864 --t2 K562_H3K27ac_treat_pileup.bdg --c2 K562_H3K27ac_control_lambda.bdg --d2 18559220 --outdir . -o H3K27ac_GM12878.bed H3K27ac_K562.bed nondiff_peak.bed 1>bdgdiff.log 2>&1 &

# H3K36me3
# GM12878 control 6284864
# K562 control 18559220
nohup macs2 bdgdiff --t1 GM12878_H3K36me3_treat_pileup.bdg --c1 GM12878_H3K36me3_control_lambda.bdg --d1 6284864 --t2 K562_H3K36me3_treat_pileup.bdg --c2 H3K36me3/K562_H3K36me3_control_lambda.bdg --d2 18559220 --outdir . -o H3K36me3_GM12878.bed H3K36me3_K562.bed nondiff_peak.bed 1>bdgdiff.log 2>&1 &

# subset data
grep '^chr6[[:space:]]' GM12878_H3K4me1_peaks.narrowPeak > GM12878_H3K4me1_peaks_chr6.narrowPeak
grep '^chr6[[:space:]]' GM12878_H3K4me3_peaks.narrowPeak > GM12878_H3K4me3_peaks_chr6.narrowPeak
grep '^chr6[[:space:]]' GM12878_H3K27ac_peaks.narrowPeak > GM12878_H3K27ac_peaks_chr6.narrowPeak
grep '^chr6[[:space:]]' GM12878_H3K36me3_peaks.narrowPeak > GM12878_H3K36me3_peaks_chr6.narrowPeak
grep '^chr6[[:space:]]' K562_H3K4me1_peaks.narrowPeak > K562_H3K4me1_peaks_chr6.narrowPeak
grep '^chr6[[:space:]]' K562_H3K4me3_peaks.narrowPeak > K562_H3K4me3_peaks_chr6.narrowPeak
grep '^chr6[[:space:]]' K562_H3K27ac_peaks.narrowPeak > K562_H3K27ac_peaks_chr6.narrowPeak
grep '^chr6[[:space:]]' K562_H3K36me3_peaks.narrowPeak > K562_H3K36me3_peaks_chr6.narrowPeak
```

Read ChIP-seq data
```{r, eval=F}
library(ChIPseeker)
K562_peak_files <- c(H3K27ac = "K562_H3K27ac_peaks_chr6.narrowPeak",
                     H3K36me3 = "K562_H3K36me3_peaks_chr6.narrowPeak",
                     H3K4me1 = "K562_H3K4me1_peaks_chr6.narrowPeak",
                     H3K4me3 = "K562_H3K4me3_peaks_chr6.narrowPeak")
epigenetics_data[["K562_peak"]] <- lapply(K562_peak_files, readPeakFile)

GM12878_peak_files <- c(H3K27ac = "GM12878_H3K27ac_peaks_chr6.narrowPeak",
                        H3K36me3 = "GM12878_H3K36me3_peaks_chr6.narrowPeak",
                        H3K4me1 = "GM12878_H3K4me1_peaks_chr6.narrowPeak",
                        H3K4me3 = "GM12878_H3K4me3_peaks_chr6.narrowPeak")
epigenetics_data[["GM12878_peak"]] <- lapply(GM12878_peak_files, readPeakFile)

H3K4me1_files <- c(peak_in_GM12878 = "H3K4me1_GM12878_chr6.bed",
                   peak_in_K562 = "H3K4me1_K562_chr6.bed")
epigenetics_data[["H3K4me1_peak"]] <- lapply(H3K4me1_files, readPeakFile)

H3K36me3_files <- c(peak_in_GM12878 = "H3K36me3_GM12878_chr6.bed",
                    peak_in_K562 = "H3K36me3_K562_chr6.bed")
epigenetics_data[["H3K36me3_peak"]] <- lapply(H3K36me3_files, readPeakFile)

H3K27ac_files <- c(peak_in_GM12878 = "./data/H3K27ac_GM12878_chr6.bed",
                   peak_in_K562 = "./data/H3K27ac_K562_chr6.bed")
epigenetics_data[["H3K27ac_peak"]] <- lapply(H3K27ac_files, readPeakFile)

H3K4me3_files <- c(peak_in_GM12878 = "./data/H3K4me3_GM12878_chr6.bed",
                   peak_in_K562 = "./data/H3K4me3_K562_chr6.bed")
epigenetics_data[["H3K4me3_peak"]] <- lapply(H3K4me3_files, readPeakFile)
```

load package

```{r}
library(ggplot2)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(aplot)
library(ChIPseeker)
library(clusterProfiler)
library(org.Hs.eg.db)
library(IRanges)
library(GenomicFeatures)
library(gggenes)

scale <- "log10"
limits <- c(0,1.7)
chr <- "chr6"
chr_hic <- "6"
x_min <- 43600000
x_max <- 44100000
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
epigenetics_data <- qs::qread("./data/epigenetics_data.qs")

# functions to visualize gene track
getPlotGene <- function(txdb, chr, x_min, x_max, tx_id = NULL,
                        OrgDb = NULL){

    win <- GRanges(seqnames = chr,
                   ranges = IRanges::IRanges(start = x_min, end = x_max))
    all_genes <- GenomicFeatures::genes(txdb)
    gene_df <- data.frame(subsetByOverlaps(x = all_genes, ranges = win, type = "any"))
    gene_df[gene_df[, "start"] < x_min, "start"] <- x_min
    gene_df[gene_df[, "end"] > x_max, "end"] <- x_max
    gene_df$gene_id <- factor(gene_df$gene_id, levels = unique(gene_df$gene_id))
    
    gene_df <- gene_df[,c("seqnames", "gene_id", "start", "end", "strand")]
    colnames(gene_df)[1] <- "chromosome"
    gene_df$forward <- ifelse(gene_df$strand=="+", TRUE, FALSE)

    if(!is.null(OrgDb)){
      require(clusterProfiler)
      require(OrgDb , character.only = TRUE)

      changeid <- bitr(geneID = gene_df$gene_id, 
                       fromType = "ENTREZID", toType = "SYMBOL",
                       OrgDb = OrgDb)
      colnames(changeid)[1] <- c("gene_id") 
      gene_df <- merge(gene_df, changeid)
      gene_df <- gene_df[, c(2:ncol(gene_df))]
      colnames(gene_df)[ncol(gene_df)] <- "gene_symbol"
    }

    return(gene_df)
}


plotGene <- function(df,x_min = NULL, x_max = NULL){
  
  require(gggenes)
  if(is.null(x_min)){
    x_min <- min(c(df[, "start"], df[, "end"]))
  }

  if(is.null(x_max)){
    x_max <- max(c(df[, "start"], df[, "end"]))
  }

  gene_p <- ggplot(df, aes(xmin = start, xmax = end, 
                           y = gene_symbol, forward=forward, fill = gene_symbol)) +
                    geom_gene_arrow() +
                    xlim(c(x_min, x_max))+
                    theme_genes() +
                    theme(legend.position = 'none')
  return(gene_p)
}
```

### combine data
```{r combine_data, fig.width=10, fig.height=15}
combine_hic_p <- HiContacts::plotMatrix(epigenetics_data[["div_hic"]], 
                                        use.scores = "balanced.fc",
                                        maxDistance = x_max - x_min,caption = FALSE,
                                        cmap = HiContacts::bbrColors()) + 
                        xlim(x_min, x_max) +
                        ylab("K562/GM12878 hic compare ") + xlab("") + labs(fill = "log2fc") + 
                        theme(plot.title = element_text(hjust = 0.5),
                              axis.text.y = element_blank(),
                              axis.ticks.y = element_blank(),
                              axis.text.x = element_text(size = 9))

peak_color <- c("#E05B5B","#519CBA")
names(peak_color) <- c("peak_in_K562", "peak_in_GM12878")

H3K27ac_p <- covplot(epigenetics_data[["H3K27ac_peak"]], 
                     weightCol="V5", chrs=chr, xlim=c(x_min, x_max),
                     ylab = "H3K27ac", title = "", xlab = "") + 
                annotate("rect", xmin = 43737946,xmax = 43754223,
                         ymin= 0, ymax=100, fill = "#c6c3c3", alpha= 0.2) + 
                annotate("rect", xmin = 43900000,xmax = 44000000,
                         ymin= 0, ymax=100, fill = "#c6c3c3" , alpha= 0.2) + 
                annotate("rect", xmin = 44060000,xmax = 44100000,
                         ymin= 0, ymax=100, fill = "#c6c3c3" , alpha= 0.2) + 
                scale_color_manual(values = peak_color) +
                scale_fill_manual(values = peak_color) +
                theme(strip.text.y = element_blank())


H3K36me3_p <- covplot(epigenetics_data[["H3K36me3_peak"]], 
                      weightCol="V5", chrs=chr, xlim=c(x_min, x_max),
                      ylab = "H3K36me3", title = "", xlab = "") + 
                annotate("rect", xmin = 43737946,xmax = 43754223,
                         ymin= 0, ymax=10, fill =  "#c6c3c3" , alpha= 0.2) + 
                annotate("rect", xmin = 43900000,xmax = 44000000,
                         ymin= 0, ymax=10, fill =  "#c6c3c3" , alpha= 0.2) + 
                annotate("rect", xmin = 44060000,xmax = 44100000,
                         ymin= 0, ymax=10,
                         fill =  "#c6c3c3" , alpha= 0.2) + 
                scale_color_manual(values = peak_color) +
                scale_fill_manual(values = peak_color) +
                theme(strip.text.y = element_blank())


H3K4me1_p <- covplot(epigenetics_data[["H3K4me1_peak"]], 
                     weightCol="V5", chrs=chr, xlim=c(x_min, x_max),
                     ylab = "H3K4me1", title = "", xlab = "") + 
                annotate("rect", xmin = 43737946,xmax = 43754223,
                         ymin= 0, ymax=30, fill =  "#c6c3c3" , alpha= 0.2) + 
                annotate("rect", xmin = 43900000,xmax = 44000000,
                         ymin= 0, ymax=30, fill =  "#c6c3c3" , alpha= 0.2) + 
                annotate("rect", xmin = 44060000,xmax = 44100000,
                         ymin= 0, ymax=30, fill =  "#c6c3c3" , alpha= 0.2) + 
                scale_color_manual(values = peak_color) +
                scale_fill_manual(values = peak_color) +
                theme(strip.text.y = element_blank())


H3K4me3_p <- covplot(epigenetics_data[["H3K4me3_peak"]], 
                     weightCol="V5", chrs=chr, xlim=c(x_min, x_max),
                     ylab = "H3K4me3", title = "", xlab = "") + 
                annotate("rect", xmin = 43737946,xmax = 43754223,
                         ymin= 0, ymax=25, fill =  "#c6c3c3" , alpha= 0.2) + 
                annotate("rect", xmin = 43900000,xmax = 44000000,
                         ymin= 0, ymax=25, fill =  "#c6c3c3" , alpha= 0.2) + 
                annotate("rect", xmin = 44060000,xmax = 44100000,
                         ymin= 0, ymax=25, fill =  "#c6c3c3" , alpha= 0.2) + 
                scale_color_manual(values = peak_color) +
                scale_fill_manual(values = peak_color) +
                theme(strip.text.y = element_blank())

gene_p <- plotGene(df = getPlotGene(txdb = txdb, 
                                    chr = chr, 
                                    x_min = x_min, 
                                    x_max = x_max,OrgDb = "org.Hs.eg.db"),
                    x_min = x_min,
                    x_max = x_max) + 
                    scale_fill_brewer(palette = "Set3") + 
                    labs(fill = "gene symbol") + 
                        annotate("rect",
                                 xmin = 43737946,xmax = 43754223,
                                 ymin= 0, ymax=9,
                                 fill =  "#c6c3c3" , alpha= 0.2) + 
                        annotate("rect",
                                 xmin = 43900000,xmax = 44000000,
                                 ymin= 0, ymax=9,
                                 fill =  "#c6c3c3" , alpha= 0.2) + 
                        annotate("rect",
                                 xmin = 44060000,xmax = 44100000,
                                 ymin= 0, ymax=9,
                                 fill =  "#c6c3c3" , alpha= 0.2)

p <- combine_hic_p %>% 
            insert_bottom(H3K27ac_p,height = .2) %>% 
            insert_bottom(H3K36me3_p,height = .2) %>% 
            insert_bottom(H3K4me1_p,height = .2) %>%
            insert_bottom(H3K4me3_p,height = .2) %>%
        insert_bottom(gene_p, height = .5)

p
```

### split data
```{r split_data, fig.width=16, fig.height=10}
gene_p <- plotGene(df = getPlotGene(txdb = txdb, 
                                    chr = chr, 
                                    x_min = x_min, 
                                    x_max = x_max,OrgDb = "org.Hs.eg.db"),
                    x_min = x_min,
                    x_max = x_max) + 
                    scale_fill_brewer(palette = "Set3") + 
                    labs(fill = "gene symbol")+ 
                  annotate("rect", xmin = 43737946,xmax = 43754223,
                           ymin= 0, ymax=9, fill =  "#c6c3c3" , alpha= 0.2) + 
                  annotate("rect", xmin = 43900000,xmax = 44000000,
                           ymin= 0, ymax=9, fill =  "#c6c3c3" , alpha= 0.2) + 
                  annotate("rect", xmin = 44060000,xmax = 44100000,
                           ymin= 0, ymax=9, fill =  "#c6c3c3" , alpha= 0.2)


GM12878_hic_p <- HiContacts::plotMatrix(epigenetics_data[["GM12878_hic"]], 
                                        maxDistance = x_max - x_min,
                                        caption = FALSE, 
                                        scale = scale, limits = limits) + 
                        xlim(x_min, x_max) + labs(fill = 'hic score') +
                        ylab("GM12878 hic") + xlab("") + 
                        theme(plot.title = element_text(hjust = 0.5),
                              axis.text.y = element_blank(),
                              axis.ticks.y = element_blank(),
                              axis.text.x = element_text(size = 9))


GM12878_cov_p <- covplot(epigenetics_data[["GM12878_peak"]], 
                         weightCol="V5", chrs=chr, xlim=c(x_min, x_max),
                         ylab = "GM12878 ChIP-seq", title = "", xlab = "") + 
                        ylim(c(0,2000)) +
                        annotate("rect", xmin = 43737946,xmax = 43754223,
                                 ymin= 0, ymax=2000, fill =  "#c6c3c3" , alpha= 0.2) + 
                        annotate("rect", xmin = 43900000,xmax = 44000000,
                                 ymin= 0, ymax=2000, fill =  "#c6c3c3" , alpha= 0.2) + 
                        annotate("rect", xmin = 44060000,xmax = 44100000,
                                 ymin= 0, ymax=2000, fill =  "#c6c3c3" , alpha= 0.2) + 
                        facet_grid(.id ~ .) +  
                        scale_fill_brewer(palette = "Set2") +
                        scale_color_brewer(palette = "Set2") + 
                        theme(strip.text.y = element_blank())
                        

GM12878_p <- (GM12878_hic_p + theme(legend.position = 'none'))  %>% 
             insert_bottom(GM12878_cov_p + theme(legend.position = 'none')) %>% 
             insert_bottom(gene_p + theme(legend.position = 'none'), height = .6)


K562_hic_p <- HiContacts::plotMatrix(epigenetics_data[["K562_hic"]], 
                                     maxDistance = x_max - x_min,
                                     caption = FALSE, scale = scale, 
                                     limits = limits) + 
                        xlim(x_min, x_max) +
                        ylab("K562 hic") + xlab("") + labs(fill = 'hic score') +
                        theme(plot.title = element_text(hjust = 0.5),
                              axis.text.y = element_blank(),
                              axis.ticks.y = element_blank(),
                              axis.text.x = element_text(size = 9))


K562_cov_p <- covplot(epigenetics_data[["K562_peak"]], 
                      weightCol="V5", chrs=chr, xlim=c(x_min, x_max),
                      ylab = "K562 ChIP-seq", title = "", xlab = "") + 
                  ylim(c(0,2000)) + 
                  annotate("rect", xmin = 43737946,xmax = 43754223,
                           ymin= 0, ymax=2000, fill =  "#c6c3c3" , alpha= 0.2) + 
                  annotate("rect", xmin = 43900000,xmax = 44000000,
                           ymin= 0, ymax=2000, fill =  "#c6c3c3" , alpha= 0.2) + 
                  annotate("rect", xmin = 44060000,xmax = 44100000,
                           ymin= 0, ymax=2000, fill =  "#c6c3c3" , alpha= 0.2) +  
                  facet_grid(.id ~ .) + 
                  scale_fill_brewer(palette = "Set2") +
                  scale_color_brewer(palette = "Set2") +  
                  theme(strip.text.y = element_blank())

K562_p <- (K562_hic_p + theme(legend.position = 'none')) %>% 
             insert_bottom(K562_cov_p + theme(legend.position = 'none')) %>% 
             insert_bottom(gene_p + theme(legend.position = 'none'), height = .6)

hic_legend <- cowplot::get_legend(K562_hic_p)
cov_legend <- cowplot::get_legend(K562_cov_p)

combined_legends <- cowplot::plot_grid(hic_legend, cov_legend, nrow = 1)
p <- plot_list(gglist = list(GM12878_p, K562_p, combined_legends),
               design = c("AAAAAABBBBBBCC"))

p
```